app/src/main/kotlin/de/markusfisch/android/binaryeye/fragment/HistoryFragment.kt

```diff
@@ -168,6 +183,63 @@ class HistoryFragment : Fragment() {
 			.show()
 	}
 
+	private fun askToExportToFile(context: Context) = scope.launch {
+		progressView.useVisibility {
+			if (!hasWritePermission(activity)) return@useVisibility
+			val (getBinaries, scans) = getScans {
+				alertDialog(context) { resume ->
+					setTitle(R.string.csv_allow_binary)
+					setMessage(R.string.csv_allow_binary_message)
+					setPositiveButton(R.string.no) { _, _ ->
+						resume(false)
+					}
+					setNegativeButton(android.R.string.yes) { _, _ ->
+						resume(true)
+					}
+					setNeutralButton(android.R.string.cancel) { _, _ ->
+						resume(null)
+					}
+				}
+			} ?: return@useVisibility
+			val delimiters = context.resources.getStringArray(
+				R.array.csv_delimiters_values
+			)
+			val delimiter = alertDialog<String>(context) { resume ->
+				setTitle(R.string.csv_delimiter)
+				setItems(R.array.csv_delimiters_names) { _, which ->
+					resume(delimiters[which])
+				}
+			} ?: return@useVisibility
+			val name = withContext(Dispatchers.Main) { activity.askForFileName(suffix = "csv") } ?: return@useVisibility
+			val csv = scans.toCSV(delimiter, getBinaries)
+			val toastMessage = saveByteArray(name, csv)
+			if (toastMessage > 0) {
+				withContext(Dispatchers.Main) {
+					Toast.makeText(
+						context,
+						toastMessage,
+						Toast.LENGTH_SHORT
+					).show()
+				}
+			}
+		}
+	}
+
+	private fun List<DatabaseRepository.Scan>.toCSV(delimiter: String, allowBinary: Boolean): ByteArray {
+		val result = mutableListOf<Byte>()
+
+		val header = "DATE${delimiter}TYPE${delimiter}CONTENT${if (allowBinary) "${delimiter}BINARY_CONTENT" else ""}\n".toByteArray()
```

Comment:
I'm not really happy with this method of creating a csv, but I don't have any good idea on how to improve it. There are 2 locations in the code where the order is defined. As you can see from the last [commit](0617695) it's bad for maintainability, as a code change requires an other change in an other place without compiler error. Any ideas on improvement?
I could refactor, to just have one function, but that's not as reusable and creates a big unclean/unreadable function.

Reply:
Found a way: ffe92eb